proptest-macro/src/property_test/validate.rs

```diff
@@ -26,34 +29,96 @@ fn all_args_non_self(f: &ItemFn) -> Result<(), TokenStream> {
     }
 }
 
+/// Make sure we only have `#[strategy = <expr>]` attributes on function parameters
+fn validate_parameter_attrs(f: &mut ItemFn) -> Result<(), TokenStream> {
+    let mut error = quote::quote! {};
+
+    for param in &mut f.sig.inputs {
+        let FnArg::Typed(pat_ty) = param else {
+            unreachable!("should be impossible due to `all_args_non_self`");
+        };
+
+        // add error for any non-`strategy` error or inner attributes (i.e. `#![...]` )
+        for attr in pat_ty.attrs.iter().filter(|a| !is_strategy(a)) {
+            error.extend(quote_spanned! {
+                attr.span() => compile_error!("only `#[strategy = <expr>]` attributes are allowed here");
+            });
+        }
+
+        let mut first_strategy_seen = false;
+        let mut final_attrs = Vec::with_capacity(pat_ty.attrs.len());
+        let old_attrs = std::mem::take(&mut pat_ty.attrs);
+
+        // every strategy attr should have the form `#[strategy = <expr>]`
+        for attr in old_attrs.into_iter().filter(is_strategy) {
+            match attr.meta {
+                // a "good" strategy - if we see more than one, emit an error
+                Meta::NameValue(_) => {
+                    if first_strategy_seen {
+                        error.extend(quote_spanned! {
+                            attr.span() => compile_error!("duplicate `#[strategy = ...]` definition");
```

Comment:
may be worth saying the name of the parameter here for added clarity

Reply:
Worth mentioning that, since we're using `quote_spanned`, the compile error will appear in IDEs to be "coming from" the second attribute (i.e. that's where the red squiggly line will appear)